enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,123 @@
+---
+title: vrf-metaplugin-cni
+authors:
+  - "@fedepaol"
+reviewers:
+  - "@dougbtv"
+  - "@zshi-redhat"
+approvers:
+  - TBD
+creation-date: 2020-07-21
+last-updated: 2020-07-21
+status: implementable
+---
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+To allow isolation, and make it possible to have overlaps between the CIDRs of secondary networks (and between them and the pod's address space), we want to introduce a VRF meta cni plugin. Taking advantage of the chaining mechanism, it can be used to assign a secondary interface to a custom [VRF](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking).
+
+## Motivation
+
+Telco customers maintain strict separation of multiple networks for different services or functionalities. When a device connects to multiple networks, the practice is to maintain strict separation of these networks.
+
+Moreover, the multiple networks a pod is connected to may have overlapping CIDRs, and those CIDRs can also overlap with the sdn’s CIDR. To maintain this kind of isolation, they typically use VRFs. Currently this is done by granting the containers the NET_ADMIN capabilities.
+
+### Goals
+
+Being able to assign a secondary interface to a particular vrf, and eventually having more secondary interfaces of the same pod assigned to the same VRF.
+
+### Non-Goals
+
+Assigning host level interfaces to VRFs.
+
+## Proposal
+
+The proposal is to introduce an additional CNI plugin that takes the vrf name as an input, and can be chained to the “main” cni used for the secondary interface, and that has the effect of creating the requested vrf (if it does not exist) and to add the interface to that vrf.
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` CNI config of `NetworkAttachmentDefinition` would look like
+
+```json
+{
+    "cniVersion": "0.3.1",
+    "name": "macvlan-vrf",
+    "plugins": [
+      {
+        "type": "macvlan",
```

Comment:
This uses `macvlan` CNI as an example, is there other CNI plugins that will be supported to use vrf chained plugin? might be worth mentioning all of them if we're not going to support all CNI types to use this vrf chained plugin.

Reply:
I'll explicitly add an example about SR-IOV. In theory, I'd expect this not depend on the specific CNI type, but those we care the most are sr-iov and macvlan.

Reply:
Yeah, I think an example with sriov CNI would be helpful. we just need to make sure there is a definite list of CNIs that will be supported to use vrf plugin and be documented. for example, it could be that we support vrf with all supported CNIs or we support vrf with macvlan & sriov CNI plugins.

Reply:
Switched to NetworkAttachmentDefinition, it's more clear in case of SR-IOV